core/src/main/java/com/linecorp/armeria/common/logging/DefaultLogWriter.java

```diff
@@ -0,0 +1,151 @@
+
+    private static boolean warnedNullRequestLogLevelMapper;
+    private static boolean warnedNullResponseLogLevelMapper;
+
+    private final Logger logger;
+    private final RequestLogLevelMapper requestLogLevelMapper;
+    private final ResponseLogLevelMapper responseLogLevelMapper;
+    private final Predicate<Throwable> responseCauseFilter;
+    private final LogFormatter logFormatter;
+
+    DefaultLogWriter(Logger logger, RequestLogLevelMapper requestLogLevelMapper,
+                     ResponseLogLevelMapper responseLogLevelMapper,
+                     Predicate<Throwable> responseCauseFilter, LogFormatter logFormatter) {
+        this.logger = logger;
+        this.requestLogLevelMapper = requestLogLevelMapper;
+        this.responseLogLevelMapper = responseLogLevelMapper;
+        this.responseCauseFilter = responseCauseFilter;
+        this.logFormatter = logFormatter;
+    }
+
+    @Override
+    public void logRequest(RequestOnlyLog log) {
+        requireNonNull(log, "log");
+        LogLevel requestLogLevel = requestLogLevelMapper.apply(log);
+        if (requestLogLevel == null) {
+            if (!warnedNullRequestLogLevelMapper) {
+                warnedNullRequestLogLevelMapper = true;
+                logger.warn("requestLogLevelMapper.apply() returned null; using default log level");
+            }
+            requestLogLevel = DEFAULT_REQUEST_LOG_LEVEL;
+        }
+        if (requestLogLevel.isEnabled(logger)) {
+            final RequestContext ctx = log.context();
+            if (log.requestCause() == null && isTransientService(ctx)) {
+                return;
+            }
+            try (SafeCloseable ignored = ctx.push()) {
+                // We don't log requestCause when it's not null because responseCause is the same exception when
+                // the requestCause is not null. That's way we don't have requestCauseSanitizer.
+                requestLogLevel.log(logger, logFormatter.formatRequest(log));
+            }
+        }
+    }
+
+    @Override
+    public void logResponse(RequestLog log) {
+        requireNonNull(log, "log");
+        LogLevel responseLogLevel = responseLogLevelMapper.apply(log);
+        if (responseLogLevel == null) {
+            if (!warnedNullResponseLogLevelMapper) {
+                warnedNullResponseLogLevelMapper = true;
+                logger.warn("responseLogLevelMapper.apply() returned null; using default log level mapper");
+            }
+            responseLogLevel = DEFAULT_RESPONSE_LOG_LEVEL_MAPPER.apply(log);
+            assert responseLogLevel != null;
+        }
+        final Throwable responseCause = log.responseCause();
+
+        if (responseLogLevel.isEnabled(logger)) {
+            final RequestContext ctx = log.context();
+            if (responseCause == null &&
+                !log.responseHeaders().status().isServerError() &&
+                isTransientService(ctx)) {
+                return;
```

Comment:
Question) Do you think it makes more sense to move this logic up to `LoggingService|Client`?
If any user decides to implement their own `LogWriter`, transient service related logic would stop working unless they consciously add this logic theirselves.

Ditto for the logic at L77

Reply:
That's a good suggestion. 👍